Proton/Sources/Core/RichTextView.swift

```diff
@@ -464,6 +464,58 @@ class RichTextView: AutogrowingTextView {
             super.toggleBoldface(sender)
         }
     }
+    
+    private func lineFragmentHeight(at textPosition: UITextPosition) -> LineHeightOffset? {
+        let index = offset(from: beginningOfDocument, to: textPosition)
+        guard layoutManager.isValidGlyphIndex(index) else {
+            // Can't get rect for the very last character. So try to use previous one instead, but ignore its origin.
+            guard let prevPosition = position(from: textPosition, offset: -1) else {
```

Comment:
why is it that both `lineFragmentHeight` and `caretRect` are looking for current and current - 1 location only?

Reply:
I’ve tried at different ranges. And every time it returns me a valid frame. The only exception is the very last position in text storage. In this case layout manager always returns `no` for `isValidGlyphIndex`. I though I could use typing attributes instead, but for text selection (not typing) it obviously doesn’t work. So I decided use frame from the previous position. It could be even previous line when the last character is empty line. So, yes, it seems like a workaround because I haven’t fount a better place to get the frame from for the very last character. 
Hope this answered the question.